### PR TITLE
Grunt support

### PR DESCRIPTION
**Pull request by crhallberg**
This adds a Gruntfile and package.json that would allow developers to use these tools. Not a replacement for the current workflow, just an alternative.

Adds:
- [x] Less compilation `grunt less`
- [x] Sass compilation `grunt sass`
- [x] JS minification `grunt js`
  - ~~eslint~~
- [x] LESS to SASS conversion
- ~~Theme packing~~ Decided to not make this PR wait for this feature
- [x] Create a wiki page for grunt
  - [x] Another page for copy-paste support for more libraries
  - [x] Link wiki page from other page(s) after merging
- [x] When merging, add package.json to list of files to update with new version number in https://vufind.org/wiki/changelog:release_checklist
- [x] Commit scss folder
- [x] Check with @lahmann (active Grunt user) to make sure we don't wreck people's existing Gruntfiles with merge conflicts.

Excited to know what else you'd like to see here!

Also would like to know if people would prefer a simple light script or one that does a lot of auto-detection out of the box. Sass or Less by default?

### CODE REVIEW

**Comment by crhallberg** [comment]
Do we want a simpler Gruntfile or more robust out of the box functionality?

**Comment by demiankatz** [comment]
I tend to favor simplicity until there is a need for complexity -- so part of the question is "who will be using this?"

I also notice that a lot of LESS files have been changed in this PR. Do these changes break the existing cssBuilder.php script? (I'm certainly open to the possibility of dropping that... but I think it would be nice if we continued to offer a solution that does not require the Node stack, if we can do so without significant extra development cost).

**Comment by crhallberg** [comment]
I'll look into keeping the LESS files the same

**Comment by demiankatz** [comment]
This PR is getting a bit overwhelming to review, and I think the main culprit may be the large number of SCSS files. Do you think it might make sense to make the SCSS stuff a second phase so that the baseline establishment of a Gruntfile is a little more readable? Or is that feature too crucial to delay?

I'm basically just trying to figure out what the minimum useful Gruntfile for the project would be. It might be nice to establish that first, because then we can treat each of the capabilities of the Gruntfile as a separate PR for separate discussion. Maybe I'm overthinking, though!

**Comment by crhallberg** [comment]
The minimal grunt is best reflected by the [LESS section](https://github.com/vufind-org/vufind/pull/659/files#diff-35b4a816e0441e6a375cd925af50752cR6) of the grunt file. You adjust the path manually and add lines for new themes.

By comparison, [this section](https://github.com/vufind-org/vufind/pull/659/files#diff-35b4a816e0441e6a375cd925af50752cR76) of the grunt file is much more robust, detecting themes, inheritance, and generating file paths automatically. Just adding a theme folder is enough to have grunt compile it.

**Comment by crhallberg** [comment]
So I made a decision to put the theme packing off so we can push this functionality forward. Theme packing development can now be found at [crhallberg/grunt-theme-pack](https://github.com/crhallberg/vufind/tree/grunt-theme-pack).

**Comment by demiankatz** [comment]
A few comments:

1.) It appears that the Grunt tasks for LESS/SASS are hard-coded to only compile the bootstrap3 and bootprint3 themes. If memory serves, the cssBuilder.php auto-detects themes and compiles all of them and/or can be told which theme to compile. I think we at least need enough flexibility to allow somebody to compile a custom sub-theme without having to modify the Gruntfile (assuming that is even possible).

2.) I notice that the version in the package.json is 3.0.0 -- we should bump that to 3.0.2 and also (after merging) note this as another file to update during the release process for new versions as documented here: https://vufind.org/wiki/changelog:release_checklist

3.) When using vendor.min.js, are we going to have problems with double-loading / double-executing in cases where vendor files are conditionally loaded by templates? It seems safe to minify all of the globally-loaded files, but the conditional ones could be problematic unless we add a mechanism to avoid loading vendor files if vendor.min.js is in place (and that seems like a relatively fragile hack).

**Comment by crhallberg** [comment]
1.) The code to perform an automated check is [fairly complex](https://gist.github.com/crhallberg/2b922eb1605413fdffebfd9b1d43c723) and I didn't feel it was stable enough. Extension is very straight-forward and [outlined on the wiki](https://vufind.org/wiki/development:grunt).

2.) Version bumped. Noted.

3.) This is a good point. We have a few options. I could manually exclude the ones that are conditionally loaded. We could grep for conditional files and exclude those programmatically. Or we can make a config that doesn't include conditional JS if set.

**Comment by demiankatz** [comment]
1.) Would it be feasible to add a command-line switch that can substitute a user-entered value for CUSTOM as outlined in the wiki? That would at least make the most common scenario where the user follows recommended VuFind best practices possible without requiring modifications to the Gruntfile. (I have no experience with Grunt, so I'm not sure how easy parameterization is... but I'm assuming there's something equivalent to Phing's -Dvariable=value mechanism).

2.) Thanks. Should we add a checkbox to the PR about the wiki update, or should we just do it now on the assumption that this will be merged soon?

3.) I would be inclined to make this as explicit as possible, at least for now. If we hard-code a list of files into the Gruntfile and put a comment in themes.config.php about commenting out vendor files and instead using the minified filename, that might cover us for now.

One other issue, though: it might be a good idea to include the name of the theme in the minified vendor filename so if we have a chain of themes with separate vendor files, they don't stomp on each other.

Another alternative is to abandon this feature for the moment. I still feel that the best solution may be to handle this on the PHP side, extending the view helper so that it serves up minified code automatically based on the scripts being requested.

Actually, here's an interesting idea:

1.) Adjust the script so that it simply takes *.js, and if the filename does not already include "min" at the end, creates a *-min.js version as a separate file.

2.) Adjust the headScript view helper so it takes a setting to collapse JS into a single file. When turned on, this takes all of the requested files, favoring any with a -min.js suffix, concatenates them into a single file, writes this to a web-accessible location, and serves that up instead of the individual parts. That will work whether or not minification has been applied, but the file will be smaller if minified versions are available. The name of the file written to the web-accessible location could be an md5 hash of all the filenames and modification dates involved. Thus, we only write the file if it does not already exist.

Does that make any sense?

Perhaps all of this suggests that this is another feature we should remove for day one so we have more time to discuss and experiment.

**Comment by crhallberg** [comment]
I've moved the automated js minification to a new PR, #757.

**Comment by crhallberg** [comment]
A command-line switch to fill in custom themes could be made, especially after simplification. This is not necessary holding up this branch, I think that is a good feature to add after this basic functionality is pushed.

**Comment by demiankatz** [comment]
Chris,

1.) I couldn't get the sass task to work -- even after running `grunt lessToSass` I ended up with a `File to import not found or unreadable: vendor/bootstrap` error in response to `grunt sass`. Are you seeing the same thing?

2.) I had an idea regarding the hard-coded themes in the Gruntfile.js -- if we stick with the convention of creating a "compiled.less" as the standard name of the primary LESS file in each theme, we can use a wildcard pattern to find all of those and compile them. I implemented this in https://github.com/vufind-org/vufind/pull/659/commits/131231be3e5b12d86c88ba563272fe1712b5e5f3 and it seems to be working correctly. However, if you dislike it, please feel free to revert. (If you agree that this is a good approach, we should update SASS to match).

3.) To test my wildcard solution, I tried compiling the vuboot3 theme from main-dev. It did compile, but the output didn't match the PHP-generated version. I suspect that this is related to your changes to how relative paths are calculated, etc. We may need to include some migration notes to help people with existing LESS files make them work with Grunt.

**Comment by crhallberg** [comment]
1.) I'll look into it tomorrow (10 Aug)

2.) I'm concerned that this will cause inheritance issues, since the order of parents needs to be included in the import path. This is what makes a magic bullet solution so difficult.

3.) I suspect this has to with the inheritance issues, I'll try it myself tomorrow.

**Comment by crhallberg** [comment]
1.) As a long as the vendor file is in place, there will be no issue on this. We are pushing sass files when this goes live, they are being left out for now to simplify the diffs.

**Comment by lahmann** [comment]
Chris, I finally found some time testing this pull request - I posted some comments and suggestions along with some code in https://github.com/crhallberg/vufind/pull/2. Let me know what you think!

File: Gruntfile.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    watch: {
+      options: {
+        atBegin: true
+      },
+      less: {
+        files: 'themes/**/*.less',
+        tasks: ['less']
+      },
+      scss: {
+        files: 'themes/**/*.scss',
```
Comment (lahmann) [inline Gruntfile.js]: Hi Chris, we just ran into an issue that grunt-watch was throwing "Maximum call stack size exceeded" warnings resulting in not compiling the scss files it's supposed to. We limited the path further down with the pattern `'themes/*/scss/**/*.scss'` as all scss files should reside in the themes scss folder and instantly got rid of the warnings. This pattern enforces the folder structure and reduces the files grunt-watch is watching - maybe something to consider for this pull request as well and the pattern could be applied to less, too.

**Comment by crhallberg** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;K1uLRcg3&#x2F;43-grunt-andre-feedback"> Grunt (Andre feedback)</a>

**Comment by demiankatz** [comment]
I've played with this some more, and it looks good (apart from the known issue that SCSS colors differ from LESS colors due to the lack of SCSS support in the accessibility library we use with LESS). With @lahmann's blessing, I think we can go ahead and merge this... but going forward, we'll need to be sure to continue to use the LESS version when compiling our standard compile.css files in order to avoid weird colors.

**Comment by lahmann** [comment]
I just tested this branch with our current master and it works pretty much out of the box - thanks a lot for all the work and adding grunt support to VuFind!

**Comment by demiankatz** [comment]
Thanks for the review/test! I'll talk to @crhallberg about merging this and finishing wiki touches when he gets in.

**Pull request merged**